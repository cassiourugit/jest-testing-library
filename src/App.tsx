import React, { useState } from 'react';

interface ListItem {
  id: number;
  text: string;
}

const App: React.FC = () => {
  const [inputText, setInputText] = useState('');
  const [listItems, setListItems] = useState<ListItem[]>([
    { id: 1, text: 'Peugeot'},
    { id: 2, text: 'Honda'},
    { id: 3, text: 'Suzuki'}
  ]);

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setInputText(e.target.value);
  };

  const handleAddItem = () => {
    if (inputText.trim() !== '') {
      const newItem: ListItem = {
        id: Date.now(),
        text: inputText
      };
      setListItems([...listItems, newItem]);
      setInputText('');
    }
  };

  const handleRemoveItem = (itemId: number) => {
    setListItems(prevItems => prevItems.filter(item => item.id !== itemId));
  };

  return (
    <div>
      <input type="text" value={inputText} onChange={handleInputChange} />
      <button onClick={handleAddItem}>Adicionar</button>
      <ul>
        {listItems.map(item => (
          <li key={item.id}>
            {item.text}
            <button onClick={() => handleRemoveItem(item.id)}>Remover</button>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default App;
