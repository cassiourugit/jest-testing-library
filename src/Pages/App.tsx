/* eslint-disable jsx-a11y/aria-role */
import React, { useEffect, useState } from 'react';
import { Button } from '../Components/Button'

interface ListItem {
  id: number;
  text: string;
}

const App: React.FC = () => {
  const [inputText, setInputText] = useState('');
  const [listItems, setListItems] = useState<ListItem[]>([
    { id: 1, text: 'Toyota'},
    { id: 2, text: 'Honda'},
    { id: 3, text: 'Suzuki'},
    { id: 4, text: 'Nissan'}
  ]);

  const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setInputText(e.target.value);
  };

  const handleAddItem = () => {
    if (inputText.trim() !== '') {
      const newItem: ListItem = {
        id: Date.now(),
        text: inputText
      };
      setListItems([...listItems, newItem]);
      setInputText('');
    }
  };

  const handleRemoveItem = (itemId: number) => {
    setListItems(prevItems => prevItems.filter(item => item.id !== itemId));
  };

  const [title, setTitle] = useState('Todos esses carros são confiáveis');

  useEffect(() => {
    const observer = new MutationObserver(() => {
      // Find the element containing the specific text
      const elementsWithText = Array.from(document.querySelectorAll('body *')).filter(element => {
        return element.textContent && element.textContent.includes('Peugeot');
      });

      // Update the title based on the presence or absence of the specific text
      if (elementsWithText.length === 0) {
        setTitle('Todos esses carros são confiáveis'); // Specific text is no longer present
      } else {
        setTitle('Nem todos esses carros são confiáveis'); // Specific text is still present
      }
    });

    // Observe changes in the DOM
    observer.observe(document.body, { childList: true, subtree: true });

    // Disconnect the observer when the component is unmounted
    return () => {
      observer.disconnect();
    };
  }, []);

  return (
    <div>
      <h2 role='title'>{title}</h2>
      <input role='input' data-testid='input_field' style={{height:35, marginRight:10, marginBottom:10}} type="text" value={inputText} onChange={handleInputChange} />
      <Button testID='btn_adicionar' onClick={handleAddItem} color={'blue'} children={'Adicionar'}></Button>
      <ul>
        {listItems.map(item => (
          <li 
          style={{marginLeft:20, marginBottom:10}} key={item.id}>
            {item.text}
              <Button testID={`btn_remover_${item.text}`} style={{marginLeft:10}} onClick={() => handleRemoveItem(item.id)} color={'red'} children={'Remover'}></Button>
          </li>
        ))}
      </ul>
    </div>
  );
};

export default App;
