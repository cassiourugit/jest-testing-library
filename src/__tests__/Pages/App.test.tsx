/* eslint-disable testing-library/no-wait-for-multiple-assertions */
/* eslint-disable no-restricted-globals */
import React from 'react'
import App from '../../Pages/App'
import { render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { create } from 'react-test-renderer'

describe('App tests', () => {
    it('Should render title', () => {
        render(<App/>)

        const title = screen.getByRole('title')

        expect(title).toHaveTextContent('Todos esses carros são confiáveis')        
    })

    it('Should render input', () => {
        render(<App/>)

        const input = screen.getByRole('input')

        expect(input).toBeTruthy()
    })

    it('Should render button Adicionar', () => {
        render(<App/>)

        const button = screen.getByText('Adicionar')

        expect(button).toBeTruthy()
    })

    it('Should render default list', () => {
        render(<App/>)

        expect(screen.getAllByRole('listitem').length).toBe(4)
        expect(screen.getByText('Toyota')).toBeTruthy()
        expect(screen.getByText('Honda')).toBeTruthy()
        expect(screen.getByText('Suzuki')).toBeTruthy()
        expect(screen.getByText('Nissan')).toBeTruthy()
    })

    it('Should add Peugeot to the list', async () => {
        render(<App/>)

        const input = screen.getByRole('input')
        const btnAdicionar = screen.getByText('Adicionar')

        userEvent.type(input, 'Peugeot')
        userEvent.click(btnAdicionar)

        await waitFor(async () => {
            expect(screen.getByText('Peugeot')).toBeTruthy()
            expect(screen.getByRole('title')).toHaveTextContent('Nem todos esses carros são confiáveis')
        })
    })

    it('Should remove Nissan from the List', async () => {
        render(<App/>)

        const btnRemover = screen.getByTestId('btn_remover_Nissan')

        userEvent.click(btnRemover)

        await waitFor(async () => {
            expect(screen.queryByText('Nissan')).toBeFalsy()
        })
    })

    it('Should render snapshot', () => {
        const { toJSON } = create(<App/>)

        expect(toJSON()).toMatchSnapshot()
    })
})