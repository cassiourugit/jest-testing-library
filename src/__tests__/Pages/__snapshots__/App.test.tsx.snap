// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App tests Should render snapshot 1`] = `
<div>
  <h2
    role="title"
  >
    Todos esses carros são confiáveis
  </h2>
  <input
    data-testid="input_field"
    onChange={[Function]}
    role="input"
    style={
      Object {
        "height": 35,
        "marginBottom": 10,
        "marginRight": 10,
      }
    }
    type="text"
    value=""
  />
  <button
    data-testid="btn_adicionar"
    onClick={[Function]}
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    style={
      Object {
        "backgroundColor": "rgb(0, 0, 255)",
        "border": "none",
        "borderRadius": "5px",
        "color": "white",
        "cursor": "pointer",
        "padding": "10px 20px",
        "transition": "background-color 0.3s ease",
      }
    }
  >
    Adicionar
  </button>
  <ul>
    <li
      style={
        Object {
          "marginBottom": 10,
          "marginLeft": 20,
        }
      }
    >
      Toyota
      <button
        data-testid="btn_remover_Toyota"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(255, 0, 0)",
            "border": "none",
            "borderRadius": "5px",
            "color": "white",
            "cursor": "pointer",
            "marginLeft": 10,
            "padding": "10px 20px",
            "transition": "background-color 0.3s ease",
          }
        }
      >
        Remover
      </button>
    </li>
    <li
      style={
        Object {
          "marginBottom": 10,
          "marginLeft": 20,
        }
      }
    >
      Honda
      <button
        data-testid="btn_remover_Honda"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(255, 0, 0)",
            "border": "none",
            "borderRadius": "5px",
            "color": "white",
            "cursor": "pointer",
            "marginLeft": 10,
            "padding": "10px 20px",
            "transition": "background-color 0.3s ease",
          }
        }
      >
        Remover
      </button>
    </li>
    <li
      style={
        Object {
          "marginBottom": 10,
          "marginLeft": 20,
        }
      }
    >
      Suzuki
      <button
        data-testid="btn_remover_Suzuki"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(255, 0, 0)",
            "border": "none",
            "borderRadius": "5px",
            "color": "white",
            "cursor": "pointer",
            "marginLeft": 10,
            "padding": "10px 20px",
            "transition": "background-color 0.3s ease",
          }
        }
      >
        Remover
      </button>
    </li>
    <li
      style={
        Object {
          "marginBottom": 10,
          "marginLeft": 20,
        }
      }
    >
      Nissan
      <button
        data-testid="btn_remover_Nissan"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        style={
          Object {
            "backgroundColor": "rgb(255, 0, 0)",
            "border": "none",
            "borderRadius": "5px",
            "color": "white",
            "cursor": "pointer",
            "marginLeft": 10,
            "padding": "10px 20px",
            "transition": "background-color 0.3s ease",
          }
        }
      >
        Remover
      </button>
    </li>
  </ul>
</div>
`;
